presto-main/src/test/java/com/facebook/presto/TestClientRequestFilterPlugin.java

```diff
@@ -0,0 +1,294 @@
+                            globallyAddedHeaders.add(headerKey);
+                        }
+                    }
+                });
+            }
+        }
+        request.setHeaders(extraHeadersMap);
+        assertEquals("CustomValue", request.getHeader("X-Custom-Header"));
+    }
+
+    private List<ClientRequestFilter> getClientRequestFilter()
+    {
+        List<ClientRequestFilter> requestFilters = new ArrayList<>();
+        ClientRequestFilter customModifier = new ClientRequestFilter()
+        {
+            @Override
+            public List<String> getHeaderNames()
+            {
+                return Collections.singletonList("X-Custom-Header");
+            }
+            @Override
+            public <T> Optional<Map<String, String>> getExtraHeaders(T additionalInfo)
+            {
+                Map<String, String> headers = new HashMap<>();
+                headers.put("X-Custom-Header", "CustomValue");
+                return Optional.of(headers);
+            }
+        };
+        requestFilters.add(customModifier);
+        return requestFilters;
+    }
+
+    @Test(
+            expectedExceptions = RuntimeException.class,
+            expectedExceptionsMessageRegExp = "Modification attempt detected: The header X-Presto-Transaction-Id is present in the blocked headers list.")
+    public void testCustomRequestFilterWithHeadersInBlockList()
+    {
+        ConcreteHttpServletRequest request = new ConcreteHttpServletRequest(ImmutableListMultimap.of("X-Custom-Header", "CustomValue1"), "http://request-modifier.com", Collections.singletonMap("attribute", "attribute1"));
+        List<ClientRequestFilter> requestFilters = getClientRequestFilterInBlockList();
+        ClientRequestFilterManager clientRequestFilterManager = TestingPrestoServer.getClientRequestFilterManager(requestFilters);
+
+        PrincipalStub testPrincipal = new PrincipalStub();
+
+        Map<String, String> extraHeadersMap = new HashMap<>();
+        Set<String> globallyAddedHeaders = new HashSet<>();
+
+        for (ClientRequestFilter requestFilter : clientRequestFilterManager.getClientRequestFilters()) {
+            boolean headersPresent = requestFilter.getHeaderNames().stream()
+                    .allMatch(headerName -> request.getHeader(headerName) != null);
+
+            if (!headersPresent) {
+                Optional<Map<String, String>> extraHeaderValueMap = requestFilter.getExtraHeaders(testPrincipal);
+
+                extraHeaderValueMap.ifPresent(map -> {
+                    for (Map.Entry<String, String> extraHeaderEntry : map.entrySet()) {
+                        String headerKey = extraHeaderEntry.getKey();
+                        if (headersBlockList.contains(headerKey)) {
+                            throw new RuntimeException("Modification attempt detected: The header " + headerKey + " is present in the blocked headers list.");
```

Comment:
Why are we duplicating code here?  I am very confused about how this test works.